From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maxim Medvedev <redrathnure@gmail.com>
Date: Sun, 2 Jul 2023 20:06:27 +0000
Subject: partial rollback of 8250 DMA patch
---
 drivers/tty/serial/8250/8250_dma.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/tty/serial/8250/8250_dma.c b/drivers/tty/serial/8250/8250_dma.c
index a442f0dfd..9634cf537 100644
--- a/drivers/tty/serial/8250/8250_dma.c
+++ b/drivers/tty/serial/8250/8250_dma.c
@@ -50,19 +50,20 @@ static void __dma_rx_complete(void *param)
 	/*
 	 * New DMA Rx can be started during the completion handler before it
 	 * could acquire port's lock and it might still be ongoing. Don't to
 	 * anything in such case.
 	 */
+	dma->rx_running = 0; // UART0 fix
 	dma_status = dmaengine_tx_status(dma->rxchan, dma->rx_cookie, &state);
-	if (dma_status == DMA_IN_PROGRESS)
-		return;
+	// UART0 fix - if (dma_status == DMA_IN_PROGRESS)
+	// UART0 fix - 	return;
 
 	count = dma->rx_size - state.residue;
 
 	tty_insert_flip_string(tty_port, dma->rx_buf, count);
 	p->port.icount.rx += count;
-	dma->rx_running = 0;
+	// UART0 fix - dma->rx_running = 0;
 
 	tty_flip_buffer_push(tty_port);
 }
 
 static void dma_rx_complete(void *param)
-- 
Created with Armbian build tools https://github.com/armbian/build

